### PR TITLE
Fixes to halide_image_io.h

### PR DESCRIPTION
**Pull request by steven-johnson**
-- JPEG case needed to use convert(), otherwise saving  non-uint8 images as JPEG produced garbage
-- improved rounding of the convert() functions (e.g. mul or div by 257 instead of 256); this may be overthinking it, but the extra overhead is unlikely to be meaningful here
-- added convert functions for bool and int64 types; these aren't likely to be very useful in practice, but it means we can instantiate compilable code for all of the supported Halide::Type C++ equivalents, and doesn't really cost us anything

### CODE REVIEW

**Comment by steven-johnson** [comment]
Monday afternoon review ping.

File: tools/halide_image_io.h [hunk 1]
```diff
@@ -43,39 +43,66 @@ inline bool CheckReturn(bool condition, const char* fmt, ...) {
...
+// code in a way that will compile. 
+
 // Convert to u8
+inline void convert(bool in, uint8_t &out) {out = -int(in);}
 inline void convert(uint8_t in, uint8_t &out) {out = in;}
-inline void convert(uint16_t in, uint8_t &out) {out = in >> 8;}
+inline void convert(uint16_t in, uint8_t &out) {
+    // We need to divide by 257 (not 256), but we can avoid division
+    // entirely using this approach (see http://research.swtch.com/divmult)
+    in = (uint32_t)(in) + (1 << 7);
```
Comment (abadams) [inline tools/halide_image_io.h]: in needs to be a 32-bit temporary

**Comment by steven-johnson** [comment]
PTAL

**Comment by abadams** [comment]
LGTM

**Pull request merged**